runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -184,17 +264,36 @@ public Object applyUpdateUnsafe(SMREntry entry, boolean isOptimistic) {
                     undoRecordFunctionMap
                             .get(entry.getSMRMethod());
             if (undoRecordTarget != null) {
-                entry.setUndoRecord(undoRecordTarget
-                        .getUndoRecord(object, entry.getSMRArguments()));
-                entry.setUndoable(true);
-                if (isOptimistic && optimisticallyUndoable) {
+                // calculate the undo record if it doesn't exist.
+                if (!entry.isUndoable()) {
+                    entry.setUndoRecord(undoRecordTarget
+                            .getUndoRecord(object, entry.getSMRArguments()));
+                    entry.setUndoable(true);
+                }
+                // If this is a standard mutation record, add it to
+                // the undo log, and enforce that the log doesn't get
+                if (!isOptimistic && needUndoRecordUnsafe()) {
+                    undoLog.addFirst(entry);
```

Comment:
testing this each time seems unnecessary. If the undoLog is not empty, then we clearly need to keep growing it. needUndoRecord() only check how far in the PAST we need to keep an undo log for.